overlord/snapstate/check_snap.go

```diff
@@ -180,13 +184,191 @@ func validateInfoAndFlags(info *snap.Info, snapst *SnapState, flags Flags) error
+	needsr := map[string]bool{
+		"meta/snap.yaml": true,
+	}
+	// needsf keeps track of things that need to be regular files (or symlinks to regular files)
+	needsf := map[string]bool{}
+	// noskipd tracks directories we want to descend into despite not being in needs*
+	noskipd := map[string]bool{}
+
+	for _, app := range s.Apps {
+		// for non-services, paths go into the needsrx bag because users
+		// need rx perms to execute it
+		bag := needsrx
+		paths := []string{app.Command}
+		if app.IsService() {
+			// services' paths just need to not be skipped by the validator
+			bag = noskipd
+			// additional paths to check for services:
+			// XXX maybe have a method on app to keep this in sync
+			paths = append(paths, app.StopCommand, app.ReloadCommand, app.PostStopCommand)
+		}
+
+		for _, path := range paths {
+			path = normPath(path)
+			if path == "" {
+				continue
+			}
+
+			needsf[path] = true
+			if app.IsService() {
+				needsx[path] = true
+			}
+			for ; path != "."; path = filepath.Dir(path) {
+				bag[path] = true
+			}
+		}
+
+		// completer is special :-/
+		if path := normPath(app.Completer); path != "" {
+			needsr[path] = true
+			for path = filepath.Dir(path); path != "."; path = filepath.Dir(path) {
+				needsrx[path] = true
+			}
+		}
+	}
+	// note all needsr so far need to be regular files
+	for k := range needsr {
+		needsf[k] = true
+	}
+	// thing can get jumbled up
+	for path := range needsrx {
+		delete(needsx, path)
+		delete(needsr, path)
+	}
+	for path := range needsx {
+		if needsr[path] {
+			delete(needsx, path)
+			delete(needsr, path)
+			needsrx[path] = true
+		}
+	}
+	seen := make(map[string]bool, len(needsx)+len(needsrx)+len(needsr))
+
+	// bad modes are logged instead of being returned because the end user
+	// can do nothing with the info (and the developer can read the logs)
+	hasBadModes := false
+	err := c.Walk(".", func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		mode := info.Mode()
+		if needsrx[path] || needsx[path] || needsr[path] {
+			seen[path] = true
+		}
+		if !needsrx[path] && !needsx[path] && !needsr[path] && !strings.HasPrefix(path, "meta/") {
+			if mode.IsDir() {
+				if noskipd[path] {
+					return nil
+				}
+				return filepath.SkipDir
+			}
+			return nil
+		}
+
+		if needsrx[path] || mode.IsDir() {
+			if mode.Perm()&0555 != 0555 {
+				logger.Noticef("in snap %q: %q should be world-readable and -executable, and isn't: %s", s.Name(), path, mode)
+				hasBadModes = true
+			}
+		} else {
+			if needsf[path] {
+				// this assumes that if it's a symlink it's OK. Arguably we
+				// should instead follow the symlink.  We'd have to expose
+				// Lstat(), and guard against loops, and ...  huge can of
+				// worms, and as this validator is meant as a developer aid
+				// more than anything else, not worth it IMHO (as I can't
+				// imagine this happening by accident).
```

Comment:
OTOH it would be nice to cover a scenario when `app.Command` symlink is dangling.

Simplistic resolving can be done like this: https://play.golang.org/p/oDwwA5x6lBo Then provide something like `snap.Container.ResolveSymlink()` which in turn would call `unsquashfs -ll <snap> <path>`, grab the exact path and the target, accumulate those in walk function as these files may still appear when walking and if they don't have a separate `WalkPaths(paths []string, walkfunc)` which calls `unsquashfs -ll <snap> <paths>`. Just a thought, though it looks like quite some work.

Reply:
the problem I don't want to address (and tried to point out in the comment) is that a symlink in a squashfs snap can look like
````
lrwxrwxrwx john/john                10 2018-01-11 08:11 ./baz -> qux -> foo -> bar
```
in `-lls`; I'd have to do two calls, one for the above (to get the target) and one to just `-ls` to get the filename, and match them up. *Much* to error prone for what it's gaining us.

Yes, it would be nicer. I don't think it's worth it (at least for now).

Reply:
I just realised I need to do something about this _anyway_. Augh.